independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java

```diff
@@ -0,0 +1,142 @@
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to read " + archive, e);
+        }
+    }
+
+    @Override
+    public <T> T processPath(String relativePath, Function<PathVisit, T> func, boolean multiReleaseSupport) {
+        if (!PathFilter.isVisible(pathFilter, relativePath)) {
+            return func.apply(null);
+        }
+        if (multiReleaseSupport) {
+            relativePath = toMultiReleaseRelativePath(relativePath);
+        }
+        try (FileSystem fs = openFs()) {
+            for (Path root : fs.getRootDirectories()) {
+                final Path path = root.resolve(relativePath);
+                if (!Files.exists(path)) {
+                    continue;
+                }
+                return PathTreeVisit.processPath(archive, root, path, pathFilter, func);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to read " + archive, e);
+        }
+        return func.apply(null);
+    }
+
+    private FileSystem openFs() throws IOException {
+        return ZipUtils.newFileSystem(archive);
+    }
+
+    @Override
+    public OpenPathTree openTree() {
+        final FileSystem fs;
+        try {
+            fs = openFs();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        synchronized (this) {
+            return new DirectoryPathTree(fs.getPath("/"), pathFilter, manifest, manifestInitialized, multiReleaseMapping) {
+
+                @Override
+                public Manifest getManifest() {
+                    final Manifest m = super.getManifest();
+                    if (!ArchivePathTree.this.manifestInitialized) {
+                        synchronized (ArchivePathTree.this) {
+                            if (!ArchivePathTree.this.manifestInitialized) {
+                                ArchivePathTree.this.manifest = m;
+                                ArchivePathTree.this.manifestInitialized = true;
+                            }
+                        }
+                    }
+                    return m;
+                }
+
+                @Override
+                public Map<String, String> getMultiReleaseMapping() {
+                    final Map<String, String> mrp = super.getMultiReleaseMapping();
+                    if (ArchivePathTree.this.multiReleaseMapping == null) {
+                        synchronized (ArchivePathTree.this) {
+                            if (ArchivePathTree.this.multiReleaseMapping == null) {
+                                ArchivePathTree.this.multiReleaseMapping = mrp;
+                            }
+                        }
+                    }
+                    return mrp;
+                }
+
+                @Override
+                public void close() throws IOException {
+                    try {
+                        super.close();
+                    } finally {
+                        fs.close();
```

Comment:
No idea how realistic either of the cases is here and maybe it was a deliberate choice, but if this fails it would shadow the exception from `super.close()` (if any).
Maybe add a `catch` and use `addSuppressed()`?